### PR TITLE
Fix `mean` bug for integral tensors.

### LINKED ISSUE #76430: [regression] int_tensor.mean(dtype=float32) should work
### 🐛 Describe the bug

https://github.com/pytorch/pytorch/pull/61643 broke the originally working case `non_fp_tensor.mean(dtype=fp_dtype)`.

Original code checks 
```
ScalarType scalarType = opt_dtype.has_value() ? opt_dtype.value() : self.scalar_type();
```
but the new code doesn't consider `opt_dtype` in checking:
```
at::isFloatingType(self_dtype) || at::isComplexType(self_dtype),
```

CC @ysiraichi who authored the PR, and @ezyang who merged it.

### Versions

1.10+

### PR DESCRIPTION
**Pull request by ysiraichi**
Fixes #76430 

**The Problem:** `opt_dtype` wasn't being taken into consideration when checking whether the input dtype was either floating point or complex dtype.

**The Solution:** run those checks with the dtype returned by `get_dtype_from_self(self, opt_dtype, true)`.

This fix restores the original behavior, before #61643. It also improves the error message so that the user better comprehends what happened. Finally, I also added 2 tests for ensuring the issue was fixed

-----
#### Before

```python
>>> a = torch.randint(0, 5, (5, 5), dtype=torch.int64)
>>> b = torch.tensor([], dtype=torch.float32)

>>> a.mean() # no dtype
RuntimeError: mean(): input dtype should be either floating point or complex dtypes. Got Long instead.

>>> a.mean(dtype=torch.float32) # with dtype
RuntimeError: mean(): input dtype should be either floating point or complex dtypes. Got Long instead.

>>> torch.mean(a, [], dtype=torch.float64, out=b) # with mismatching dtype and out dtype
RuntimeError: mean(): input dtype should be either floating point or complex dtypes. Got Long instead.
```

#### After

```python
>>> a = torch.randint(0, 5, (5, 5), dtype=torch.int64)
>>> b = torch.tensor([], dtype=torch.float32)

>>> a.mean() # no dtype
RuntimeError: mean(): at least one of (i) the input dtype and (ii) the desired output dtype should be either floating point or complex. Got (i) Long and (ii) None instead.

>>> a.mean(dtype=torch.float32) # with dtype
tensor(1.6800)

>>> torch.mean(a, [], dtype=torch.float64, out=b) # with mismatching dtype and out dtype
RuntimeError: Expected out tensor to have dtype double, but got float instead
```

### CODE REVIEW

**Comment by ysiraichi** [comment]
@ngimel [suggested casting the input](https://github.com/pytorch/pytorch/issues/76430#issuecomment-1111500175) directly to a floating point tensor. But, since I wasn't sure what the output dtype should be (`half`, `float`, or `double`), I simply fixed it so as to match the old behavior.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -209,7 +209,19 @@ void check_floating_or_complex_dtype(const char* name, ScalarType dtype) {
 
 TORCH_META_FUNC2(mean, dim)
 (const Tensor& self, IntArrayRef dim, bool keepdim, optional<ScalarType> opt_dtype) {
-  check_floating_or_complex_dtype("mean", self.scalar_type());
+  auto in_dtype = at::native::get_dtype_from_self(self, opt_dtype, true);
+
+  std::string opt_dtype_str("None");
```
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:214]: Probably want to hide this computation in a conditional that checks the error conditions so we're not creating this string on every call

File: test/test_reductions.py [hunk 2]
```diff
@@ -744,6 +744,28 @@ def test_logsumexp_dim(self, device):
             lambda n, d: logsumexp(n, d),
             use_integral=False)
 
+    @onlyCPU
+    def test_mean_int(self, device):
```
Comment (mruberry) [inline test/test_reductions.py:748]: I think we can model this as an ErrorInput for the mean OpInfo?

File: test/test_reductions.py [hunk 3]
```diff
@@ -744,6 +744,28 @@ def test_logsumexp_dim(self, device):
...
+        a = make_tensor((3, 4, 5), dtype=torch.int64, device=device)
+
+        # Can't compute mean of integral tensors.
+        with self.assertRaisesRegex(RuntimeError, r"mean\(\): at least one of .* Got \(i\) Long and \(ii\) None"):
+            a.mean()
+
+
+    @onlyCPU
+    def test_mean_int_with_optdtype(self, device):
```
Comment (mruberry) [inline test/test_reductions.py:757]: Same thing -- can we make this an ErrorInput?

**Comment by ysiraichi** [comment]
> about auto-casting integral tensor (and fixing #64897): I propose to auto-cast to `torch.get_default_dtype()`

I think that makes sense to me, and hopefully it is not too confusing. 
Given `torch.mean(t, [], dtype=opt_dtype)`:

| `t` | `opt_dtype` | Output dtype |
|-----|----------------|-------------------|
| integral tensor | `None` |  `torch.get_default_dtype()` |
| * | floating or complex dtype | `opt_dtype` |
| floating or complex tensor | `None` |  `t.dtype` |

@mruberry @ngimel 
What do you think?
That said, I think it is better to leave it to a future PR.

**Comment by mruberry** [comment]
> > about auto-casting integral tensor (and fixing #64897): I propose to auto-cast to `torch.get_default_dtype()`
> 
> I think that makes sense to me, and hopefully it is not too confusing. Given `torch.mean(t, [], dtype=opt_dtype)`:
> 
> `t`	`opt_dtype`	Output dtype
> integral tensor	`None`	`torch.get_default_dtype()`
> *	floating or complex dtype	`opt_dtype`
> floating or complex tensor	`None`	`t.dtype`
> @mruberry @ngimel What do you think? That said, I think it is better to leave it to a future PR.

That looks correct to me, but I'll let @ngimel review this more thoroughly since she's been working closely with reductions recently

**Comment by vadimkantorov** [comment]
For integral tensors one funny option may also be to do sum in int64 (or int32?) and then divide by the size - but maybe this might actually be slower and have worse dynamic range behavior than float

**Comment by ysiraichi** [comment]
@mruberry @ngimel  
This is a friendly reminder. Do you have some time to take a look at this PR?
I think it is better to introduce the changes for supporting integral tensors in a separate PR (that's why I didn't implement them here).

File: aten/src/ATen/native/ReduceOps.cpp [hunk 4]
```diff
@@ -200,16 +200,23 @@ TORCH_META_FUNC2(prod, dim_int)
...
+
+  if (!at::isFloatingType(in_dtype) && !at::isComplexType(in_dtype)) {
+    std::string opt_dtype_str("None");
+    if (opt_dtype.has_value()) {
+      opt_dtype_str = toString(opt_dtype.value());
+    }
+
+    TORCH_CHECK(
+        false,
+        "mean(): at least one of (i) the input dtype and (ii) the desired output dtype should be "
```
Comment (ngimel) [inline aten/src/ATen/native/ReduceOps.cpp:215]: this is a misleading error message, e.g. if self dtype if float but dtype arg is integer, then it's and invalid combination (and you'll rightly error out), but error message wouldn't explain it.

**Comment by ysiraichi** [comment]
@mruberry @ngimel 
This is a friendly reminder. Do you have some time to take a look at this PR?

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**